### ISSUE TITLE
linter: no-redeclare reports TypeScript value/type same-name declarations

### ISSUE THREAD

**Issue by maro114510** [oxc-project/oxc#25906]
### What version of Oxlint are you using?

1.79.0

### What command did you run?

`oxlint`

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "categories": {
    "correctness": "off"
  },
  "rules": {
    "no-redeclare": "error"
  }
}
```

### What happened?

The following valid TypeScript code reports `no-redeclare`:

```ts
export const Hoge = {
  A: "A",
} as const;

export type Hoge = (typeof Hoge)[keyof typeof Hoge];
```

```text
'Hoge' is already defined.
```

This appears to be the same case as #8237, which is now closed.

I found that #25691 changed `no-redeclare` to run in ES modules while skipping only the `builtinGlobals` check, so this case is now reported in ESM TypeScript files as well.

I also noticed that #10402 added the following as an explicit failing test:

```ts
type foo = 1;
export function foo(): void;
export function foo() {}
```

So reporting TypeScript value/type name collisions may be intentional. `@typescript-eslint/no-redeclare` also reports this kind of collision, including when `ignoreDeclarationMerge` is enabled.

Is this the intended behavior for Oxlint's `eslint/no-redeclare`?

If so, it may be worth documenting this TypeScript-specific behavior, since the rule documentation currently only mentions the `builtinGlobals` option.

**Comment by camc314** [comment]
Hi! Thanks for reporting this issue.

This behaviour aligns with the upstream typescript-eslint rule.

https://typescript-eslint.io/play/#ts=6.0.2&showAST=scope&fileType=.tsx&code=FAUwHgDg9gTgLgAgMZQHYGdEBkQDcQA2CAvAgN7AIICSqAZlAFwIDkAlvVCwDSUICiMGLGYsQQ2D2ABfBAEN0yNJmChIsRHACeEEAhz4ipABTbdUOvryEAlAG0A1iC0WEZkK4OEAukA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNEy8AIY90URNGgdokcGAC%2BIeUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

I would recommend disabling this rule for typescript files as it is checked already by the typescript compiler.

Thanks!
